From 9a77f7e1ffffcc7d18e1792d93a3ed1ce7ec849a Mon Sep 17 00:00:00 2001
From: Yumi Yukimura <me.cafebabe@gmail.com>
Date: Thu, 24 Jul 2025 20:15:55 +0800
Subject: [PATCH 21/21] block_dev_initializer: Find block device name in boot
 devices as fallback

* Fixes issues like /dev/block/sda1 not appearing if it has
  uevent.partition_name and fstab wants /dev/block/sda1 but not
  by-name symlink to the partition

Change-Id: I86105f36b031ba423a7bec70ed37db0c4388f57e
---
 init/block_dev_initializer.cpp | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/init/block_dev_initializer.cpp b/init/block_dev_initializer.cpp
index deb68e9ac..4111ca21c 100644
--- a/init/block_dev_initializer.cpp
+++ b/init/block_dev_initializer.cpp
@@ -120,13 +120,12 @@ ListenerAction BlockDevInitializer::HandleUevent(const Uevent& uevent,
     }
 
     auto name = uevent.partition_name;
-    if (name.empty()) {
-        size_t base_idx = uevent.path.rfind('/');
-        if (base_idx == std::string::npos) {
-            return ListenerAction::kContinue;
-        }
-        name = uevent.path.substr(base_idx + 1);
+
+    size_t base_idx = uevent.path.rfind('/');
+    if (base_idx == std::string::npos) {
+        return ListenerAction::kContinue;
     }
+    auto block_device_name = uevent.path.substr(base_idx + 1);
 
     auto iter = devices->find(name);
     if (iter == devices->end()) {
@@ -135,7 +134,11 @@ ListenerAction BlockDevInitializer::HandleUevent(const Uevent& uevent,
             iter = devices->find(partition_name);
         }
         if (iter == devices->end()) {
-            return ListenerAction::kContinue;
+            name = block_device_name;
+            iter = devices->find(name);
+            if (iter == devices->end()) {
+                return ListenerAction::kContinue;
+            }
         }
     }
 
-- 
2.47.3

